docs/pages/indexing.md

```diff
@@ -1,18 +1,325 @@
+to the indexing operator.
+
+\snippet test/index.cpp index_tutorial_second_and_fourth_rows
+
+\f[
+A(seq(1, end, 2), span) =
+\begin{bmatrix}
+     1 & 5 &  9 & 13 \\
+     3 & 7 & 11 & 15
+\end{bmatrix}
+\f]
+
+## Indexing using af::array
+
+You can also index using other af::array objects. ArrayFire performs a Cartesian
+product of the input arrays.
+
+\snippet test/index.cpp index_tutorial_array_indexing
+
+\f[
+A =
+\begin{bmatrix}
+    0 & 4 & 8 & 12 \\
+    1 & 5 & 9 & 13 \\
+    2 & 6 & 10 & 14 \\
+    3 & 7 & 11 & 15
+\end{bmatrix}
+\\
+A(
+\begin{bmatrix}
+2 \\ 1 \\ 3
+\end{bmatrix}
+,
+\begin{bmatrix}
+3 \\ 1 \\ 2
+\end{bmatrix}
+) =
+
+\begin{bmatrix}
+(2,3) & (2,1) & (2,2) \\
+(1,3) & (1,1) & (1,2) \\
+(3,3) & (3,1) & (3,2)
+\end{bmatrix}
+=
+\begin{bmatrix}
+14 & 6 & 10 \\
+13 & 5 &  9 \\
+15 & 7 & 11
+\end{bmatrix}
+\f]
+
+
+If you want to index an af::array using coordinate arrays, you can do that using the
+af::approx1 and af::approx2 functions.
+
+\snippet test/index.cpp index_tutorial_approx
+
+\f[
+approx2(A,
+\begin{bmatrix}
+2 \\ 1 \\ 3
+\end{bmatrix}
+,
+\begin{bmatrix}
+3 \\ 1 \\ 2
+\end{bmatrix}
+) =
+\begin{bmatrix}
+(2,3) \\
+(1,1) \\
+(3,2)
+\end{bmatrix}
+=
+\begin{bmatrix}
+14 \\
+ 5 \\
+11
+\end{bmatrix}
+\f]
+
+Boolean(b8) arrays can be used to index into another array. In this type of
+indexing the non-zero values will be selected by the boolean operation. If we
+want to select all values less than 5, we can pass a boolean expression into
+the parenthesis operator.
+
+\snippet test/index.cpp index_tutorial_boolean
+
+\f[
+out =
+\begin{bmatrix}
+0 \\ 1 \\ 2 \\ 3 \\ 4
+\end{bmatrix}
+\f]
+
+## References and copies
+
+All ArrayFire indexing functions return af::array(technically its an array_proxy
+class but we will discuss that later) objects. These objects may be new arrays
+or they may reference the original array depending on the type of indexing that
+was performed on them. If an array was indexed using another af::array or it
+was indexed using the af::approx functions, then a new array is created. It does
+not reference the original data.
+
+If an array was indexed using a scalar, af::seq or af::span, then the resulting
+array will reference the original data IF the first dimension is continuous. The
+following lines will not allocate additional memory.
+
+\snippet test/index.cpp index_tutorial_references
+
+The following code snippet shows some examples of indexing that will allocate
+new memory.
+
+\snippet test/index.cpp index_tutorial_copies
+
+Notice that even though the copy3 array is referencing continuous memory in the
+original array, a new array is created because we used an array to index into
+the af::array.
+
+## Assignment
+
+An assignment on an af::array, will update the data of the array with the result
+of the expression on the right hand side of the equal(=) operator. Assignments
+will not updated the array was previously referencing through an indexing
```

Comment:
typo, `will not update` and something else is missing, perhaps a conjunction `that`.

Reply:
```
An assignment on an af::array, will replace the array with the result of the
expression on the right hand side of the equal(=) operator. This means that the
type and shape of the result can be different from the array on the left had
side of the equal operator. Assignments will not update the array that was
previously referenced through an indexing operation. Here is an example:
```